libs/shared/src/main/java/io/crate/common/collections/accountable/AccountableList.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.common.collections.accountable;
+
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.function.IntConsumer;
+
+/**
+ * List implementation accounting for shallow memory usage.
+ * Based on {@link ArrayList} with some simplifications.
+ * Doesn't have concurrent modification check.
+ * Supports boundary checks since some features (lag/lead) rely on IndexOutOfBoundsException.
+ * Supports only {@link #add(Object)} operation.
+ */
+public class AccountableList<T> extends AbstractList<T> {
+
+    private  static final int SOFT_MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private static final int DEFAULT_CAPACITY = 10;
```

Comment:
Initially I wanted to start with fixed size (accept `Object[] `as initial data) and simplify even more and remove all default capacity related code but that didn't work well when integration with `BatchIterator`-s.  

It's stream processing so there is no "initial data" (or would be really cumbersome to get it), better to start from clean state. 

I still can remove `DEFAULT_CAPACITY` and let it grow 2*x from the very beginning (in case we don't expect window functions produce short arrays). 

For now aligned it with with `ArrayList`.